src/shared/api.ts

```diff
@@ -46,14 +49,32 @@ export interface ModelInfo {
 	contextWindow?: number
 	supportsImages?: boolean
 	supportsComputerUse?: boolean
-	supportsPromptCache: boolean // this value is hardcoded for now
+	supportsPromptCache: boolean
 	inputPrice?: number
 	outputPrice?: number
 	cacheWritesPrice?: number
 	cacheReadsPrice?: number
 	description?: string
 }
 
+// DeepSeek
+export type DeepSeekModelId = keyof typeof deepSeekModels
+export const deepSeekDefaultModelId: DeepSeekModelId = "deepseek-chat"
+export const deepSeekModels = {
+	"deepseek-chat": {
+		maxTokens: 4096,
+		contextWindow: 64_000,
+		supportsImages: false,
+		supportsComputerUse: false,
+		supportsPromptCache: true,
+		inputPrice: 0.14,
+		outputPrice: 0.28,
+		cacheWritesPrice: 0.14,
+		cacheReadsPrice: 0.014,
```

Comment:
While https://api-docs.deepseek.com/news/news0802 lists a price of `0.014`, that rate applies only to cache hits.

* If a cache miss occurs, isn't the price actually `0.14`?

* Since reading from the cache does not guarantee a hit, shouldn't we pessimistically assume the higher rate of `0.14` instead of `0.014`?

To be clear — While I see `prompt_cache_hit_tokens` at https://github.com/cline/cline/blob/eeaeabca95db019ebf7253c4b8aa2519e0f2d845/src/api/providers/deepseek.ts#L70 — I don't see logic about how `prompt_cache_hit_tokens` is incorporated into the calculation here.

Therefore, I suggest we pessimistically assume the higher rate of `0.14`:

```suggestion
		cacheReadsPrice: 0.014,
```

Reply:
Thanks for the feedback!  I'm not sure if I understand correctly, but after looking at the documentation again, the current implementation with cacheReadsPrice: 0.014 is actually correct.

The DeepSeek API explicitly tells us which tokens were cache hits vs misses through two fields in the usage statistics:

- prompt_cache_hit_tokens (tokens that hit the cache)
- prompt_cache_miss_tokens (tokens that missed the cache)

The implementation maps these directly:

```
yield {
    type: "usage",
    cacheReadTokens: usage.prompt_cache_hit_tokens || 0,    // Cache hits
    cacheWriteTokens: usage.prompt_cache_miss_tokens || 0,  // Cache misses
}
```
Then the cost calculator applies:

- cacheReadsPrice (0.014) to cache hits
- cacheWritesPrice (0.14) to cache misses

We don't need to pessimistically assume the higher rate because the API explicitly tells us which tokens hit vs missed the cache, and we apply the appropriate rate to each. Using 0.14 for cache hits would actually inflate the displayed cost for tokens that successfully hit the cache.

This matches both the API's pricing documentation and how other providers like Anthropic handle cache pricing (they also use separate rates for hits vs misses).

If you feel this is wrong, please elaborate. Thank you in advance.

Reply:
Thank you for your clarification.

If I understand what you are saying, DeepSeek doesn't impose any hidden charges for cache writes that occur after a cache miss. I want to highlight this point because the pricing structures of Anthropic and DeepSeek published pricing in different ways.

Anthropic explicitly lists Cache Writes and Cache Hits at https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching#pricing. In contrast, DeepSeek mentions Cache Hits and Cache Misses at https://api-docs.deepseek.com/quick_start/pricing, but it notably omits any reference to Cache Writes. Unfortunately, DeepSeek's documentation doesn't clarify whether additional fees for cache writes following cache misses exist.

Here are multiple steps involved in the sequence of operations:

1. Cache Read.
2. Cache Hit or Miss.
3. Cache Write.
4. Hypothetically, a Cache Write may be followed by an additional Cache Read to keep the cache warm.

I'm curious whether DeepSeek charges a total price of 0.14 for both operations (Cache Misses and Cache Writes) or if the pricing is 0.14 plus an additional 0.014, resulting in a total of 0.154 for both steps.

For context, I have yet to use DeepSeek; my review of this pull request is based solely on my reading of the documentation and the differences in wording between Anthropic and DeepSeek.

As a suggestion for you as a DeepSeek customer, to resolve any ambiguity, consider contacting DeepSeek directly for clarification. You might ask:

> DeepSeek's documentation lists Cache Hits and Cache Misses but doesn't explicitly mention Cache Writes. Does the price for a Cache Miss (0.14) include the cost of the subsequent write operation, or is there an additional fee for the write (e.g., 0.14 + 0.014 = 0.154 for both steps)?

The above question can help ensure precision to avoid any misinterpretations.

Reply:
You make a good point and I appreciate your attention to this detail. 

I will reach out to them to clarify, referencing this PR.

Reply:
Thanks!

Reply:
I have received an answer from DeepSeek:

![CleanShot 2024-11-28 at 20 05 46@2x](https://github.com/user-attachments/assets/5a500f02-348c-4ae8-997e-4de58cde75ff)

Reply:
Thanks!

Reply:
Interestingly, a few weeks ago, the results with DeepSeek were ... let's say "mixed" or "unsatisfactory".
I used it again today and there's definitely a big difference in the quality of the model's response. 
It's still slow though.. 

Pictures say more than a thousand words:
800+ SLOC; it makes small changes and writes the complete file. 
![CleanShot 2024-12-13 at 03 14 02@2x](https://github.com/user-attachments/assets/c7990d41-5c81-4137-96b3-81e9e5bcf110)